### DISCUSSION TITLE
CSV to parquet with read_csv_auto(sample_size=-1) on large CSV files

### DISCUSSION THREAD

**Discussion by vinhdizzo** [duckdb/duckdb#4223] [General]
Hi,

Not an issue, but I have a question.  Apologies for leveraging this avenue as I'm not sure where is the place to pose such a question.

I plan to use `duckdb` to convert a lot of CSV files to parquet files via the `COPY` statement and the `read_csv_auto` function, something like the following:

```
COPY (SELECT * FROM read_csv_auto('/dev/stdin', delim='\t', header=True, NULLSTR='\N', SAMPLE_SIZE=-1) ) TO 'my_file.parquet' (FORMAT 'PARQUET', CODEC 'GZIP')
```

Note the `SAMPLE_SIZE=-1` option, which I assume means to read every row in the CSV in order for `duckdb` to determine the right data type for each column.  Would there be any issue or concerns when using this option, especially for very large CSV files (say, 200GB)?  Are there memory requirements for such a run on very large files?  I know a full file scan means a longer run time, but this ensures that the column types are correct.  For very large files, I've run into issues where INT32 was selected for a column type, but there are rows that require INT64.

Thank you in advance for your assistance and guidance.

**Comment by Alex-Monahan** [comment]
I am not sure how it would perform, and it depends on how much RAM you have!:-)

Another option would be to use a smaller sample size, but then query the metadata about the file to get the column types. You could then dynamically write another SQL statement to take all int32 columns and use int64 instead (dynamic SQL / writing SQL with SQL). (Maybe write the top 10000 rows into a temp table, then query the metadaya of that temp table)

Here is an example of dynamic SQL for a different purpose (and with SQLite, but the approach is similar) https://til.simonwillison.net/sqlite/column-combinations

**Comment by vinhdizzo** [comment]
@Alex-Monahan Do you know if `read_csv_auto` from DuckDB reads in the entire file into RAM in order to determine column type?  If so, then I guess my approach would not be good.  But if it doesn't, I don't mind two full pass-throughs for the conversion to parquet.

**Reply by Mytherin** [reply]
If you are reading from a socket (e.g. using `httpfs` or from `stdin`) the entire file will be read into memory. Otherwise the file will be stream read from disk twice.

**Reply by vinhdizzo** [reply]
Ahh, thanks @Mytherin for pointing out the distinction between an import from stdin vs a file on disk.

So, what I am inferring from your response is, if the CSV file is on disk, then the `sample_size=-1` option shouldn't raise any concerns regardless of the file size. Is my interpretation correct? Thank you.
